gradle 6.0-rc-1 throws NoSuchMethodError when using scala 2.13.1

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->
When trying 6.0 rc1 with scala 2.13.1 it fails with a NoSuchMethodException. Trying other versions (2.11.12, 2.12.10, and 2.13.0) it works fine. A minimal example to reproduce is available at:

https://github.com/brharrington/gradle-scala

At first glance it looks similar to a problem seen with the `scala-maven-plugin`: https://github.com/davidB/scala-maven-plugin/issues/363

### Expected Behavior
<!--- Tell us what should happen -->
Should compile and run successfully.

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
Fails with:

```
$ ./gradlew run

> Task :compileScala FAILED
java.lang.NoSuchMethodError: scala.tools.nsc.Settings.nowarn()Lscala/tools/nsc/settings/AbsSettings$AbsSetting;
xsbt.Command$.getNoWarn(Command.scala:42)
xsbt.DelegatingReporter$.apply(DelegatingReporter.scala:23)
xsbt.CachedCompiler0.<init>(CompilerInterface.scala:85)
xsbt.CompilerInterface.newCompiler(CompilerInterface.scala:28)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:248)
sbt.internal.inc.AnalyzingCompiler.newCachedCompiler(AnalyzingCompiler.scala:145)
sbt.internal.inc.AnalyzingCompiler.newCachedCompiler(AnalyzingCompiler.scala:132)
sbt.internal.inc.FreshCompilerCache.apply(CompilerCache.scala:102)
sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:91)
scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:186)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3(MixedAnalyzingCompiler.scala:82)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3$adapted(MixedAnalyzingCompiler.scala:77)
sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:215)
sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:77)
sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:146)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:343)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:343)
sbt.internal.inc.Incremental$.doCompile(Incremental.scala:120)
sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:100)
sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:180)
sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:98)
sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:102)
sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:155)
sbt.internal.inc.Incremental$.compile(Incremental.scala:92)
sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:75)
sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:348)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:301)
sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:168)
sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:248)
sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:74)
org.gradle.api.internal.tasks.scala.ZincScalaCompiler.execute(ZincScalaCompiler.java:158)
org.gradle.api.internal.tasks.scala.ZincScalaCompilerFacade.execute(ZincScalaCompilerFacade.java:38)
org.gradle.api.internal.tasks.scala.ZincScalaCompilerFacade.execute(ZincScalaCompilerFacade.java:26)
org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:113)
org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:50)
org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:47)
org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:37)
org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:98)
org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:37)
org.gradle.workers.internal.IsolatedClassloaderWorker.execute(IsolatedClassloaderWorker.java:49)
org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:84)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:129)
org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:126)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:412)
org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
java.lang.Thread.run(Thread.java:748)
           

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileScala'.
> scala.tools.nsc.Settings.nowarn()Lscala/tools/nsc/settings/AbsSettings$AbsSetting;

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
1 actionable task: 1 executed
```

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Just trying the RC.

### Steps to Reproduce 
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->
```
$ git clone https://github.com/brharrington/gradle-scala.git
$ cd gradle-scala
$ ./gradlew run
```

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:

--- a/subprojects/build-cache/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java
+++ b/subprojects/build-cache/src/main/java/org/gradle/caching/internal/origin/OriginMetadataFactory.java
@@ -24,10 +24,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
-import java.net.InetAddress;
-import java.net.UnknownHostException;
 import java.util.Properties;
-import java.util.function.Consumer;
 
 public class OriginMetadataFactory {
 
@@ -43,20 +40,27 @@ public class OriginMetadataFactory {
     private static final String HOST_NAME_KEY = "hostName";
     private static final String USER_NAME_KEY = "userName";
 
+    private final File rootDir;
     private final String userName;
     private final String operatingSystem;
     private final String currentBuildInvocationId;
-    private final Consumer<Properties> additionalProperties;
-    private final File rootDir;
+    private final PropertiesConfigurator additionalProperties;
+    private final HostnameLookup hostnameLookup;
 
-    private volatile String localHostName;
-
-    public OriginMetadataFactory(File rootDir, String userName, String operatingSystem, String currentBuildInvocationId, Consumer<Properties> additionalProperties) {
+    public OriginMetadataFactory(
+        File rootDir,
+        String userName,
+        String operatingSystem,
+        String currentBuildInvocationId,
+        PropertiesConfigurator additionalProperties,
+        HostnameLookup hostnameLookup
+    ) {
         this.rootDir = rootDir;
         this.userName = userName;
         this.operatingSystem = operatingSystem;
         this.additionalProperties = additionalProperties;
         this.currentBuildInvocationId = currentBuildInvocationId;
+        this.hostnameLookup = hostnameLookup;
     }
 
     public OriginWriter createWriter(CacheableEntity entry, long elapsedTime) {
@@ -71,23 +75,12 @@ public class OriginMetadataFactory {
                 properties.setProperty(EXECUTION_TIME_KEY, Long.toString(elapsedTime));
                 properties.setProperty(ROOT_PATH_KEY, rootDir.getAbsolutePath());
                 properties.setProperty(OPERATING_SYSTEM_KEY, operatingSystem);
-                properties.setProperty(HOST_NAME_KEY, determineHostName());
+                properties.setProperty(HOST_NAME_KEY, hostnameLookup.getHostname());
                 properties.setProperty(USER_NAME_KEY, userName);
-                additionalProperties.accept(properties);
+                additionalProperties.configure(properties);
                 properties.store(outputStream, "Generated origin information");
             }
         };
-    }
-
-    private String determineHostName() {
-        if (localHostName == null) {
-            try {
-                localHostName = InetAddress.getLocalHost().getHostName();
-            } catch (UnknownHostException e) {
-                localHostName = "<unknown>";
-            }
-        }
-        return localHostName;
     }
 
     public OriginReader createReader(CacheableEntity entry) {
@@ -111,5 +104,13 @@ public class OriginMetadataFactory {
                 return new OriginMetadata(originBuildInvocationId, originalExecutionTime);
             }
         };
+    }
+
+    public interface PropertiesConfigurator {
+        void configure(Properties properties);
+    }
+
+    public interface HostnameLookup {
+        String getHostname();
     }
 }

--- a/subprojects/build-cache/src/test/groovy/org/gradle/caching/internal/origin/OriginMetadataFactoryTest.groovy
+++ b/subprojects/build-cache/src/test/groovy/org/gradle/caching/internal/origin/OriginMetadataFactoryTest.groovy
@@ -23,7 +23,14 @@ class OriginMetadataFactoryTest extends Specification {
     def entry = Mock(CacheableEntity)
     def rootDir = Mock(File)
     def buildInvocationId = UUID.randomUUID().toString()
-    def factory = new OriginMetadataFactory(rootDir, "user", "os", buildInvocationId, { it.gradleVersion = "3.0" })
+    def factory = new OriginMetadataFactory(
+        rootDir,
+        "user",
+        "os",
+        buildInvocationId,
+        { it.gradleVersion = "3.0" },
+        { "my-host" }
+    )
 
     def "converts to origin metadata"() {
         entry.identity >> "identity"
@@ -46,7 +53,7 @@ class OriginMetadataFactoryTest extends Specification {
         origin.executionTime == "10"
         origin.rootPath == "root"
         origin.operatingSystem == "os"
-        origin.hostName == InetAddress.localHost.hostName
+        origin.hostName == "my-host"
         origin.userName == "user"
         origin.buildInvocationId == buildInvocationId
     }

--- a/subprojects/core/src/main/java/org/gradle/caching/internal/services/BuildCacheServices.java
+++ b/subprojects/core/src/main/java/org/gradle/caching/internal/services/BuildCacheServices.java
@@ -37,6 +37,7 @@ import org.gradle.internal.instantiation.InstantiatorFactory;
 import org.gradle.internal.nativeintegration.filesystem.FileSystem;
 import org.gradle.internal.operations.BuildOperationExecutor;
 import org.gradle.internal.os.OperatingSystem;
+import org.gradle.internal.remote.internal.inet.InetAddressFactory;
 import org.gradle.internal.scopeids.id.BuildInvocationScopeId;
 import org.gradle.internal.service.ServiceRegistry;
 import org.gradle.internal.snapshot.FileSystemMirror;
@@ -65,8 +66,9 @@ public class BuildCacheServices {
     }
 
     OriginMetadataFactory createOriginMetadataFactory(
+        BuildInvocationScopeId buildInvocationScopeId,
         GradleInternal gradleInternal,
-        BuildInvocationScopeId buildInvocationScopeId
+        InetAddressFactory inetAddressFactory
     ) {
         File rootDir = gradleInternal.getRootProject().getRootDir();
         return new OriginMetadataFactory(
@@ -76,7 +78,8 @@ public class BuildCacheServices {
             buildInvocationScopeId.getId().asString(),
             properties -> {
                 properties.setProperty(GRADLE_VERSION_KEY, GradleVersion.current().getVersion());
-            }
+            },
+            inetAddressFactory::getHostname
         );
     }
 

--- a/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/InetAddressFactory.java
+++ b/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/InetAddressFactory.java
@@ -18,6 +18,7 @@ package org.gradle.internal.remote.internal.inet;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.annotation.Nullable;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.UnknownHostException;
@@ -141,7 +142,7 @@ public class InetAddressFactory {
         }
     }
 
-
+    @Nullable
     private InetAddress findOpenshiftAddresses() {
         for (String key : System.getenv().keySet()) {
             if (key.startsWith("OPENSHIFT_") && key.endsWith("_IP")) {

--- a/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/InetAddressFactory.java
+++ b/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/InetAddressFactory.java
@@ -15,10 +15,13 @@
  */
 package org.gradle.internal.remote.internal.inet;
 
+import org.gradle.internal.os.OperatingSystem;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.annotation.Nullable;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.UnknownHostException;
@@ -39,16 +42,41 @@ public class InetAddressFactory {
     private String hostName;
 
     public String getHostname() {
-        synchronized (lock) {
-            if (hostName == null) {
-                try {
-                    hostName = InetAddress.getLocalHost().getHostName();
-                } catch (UnknownHostException e) {
-                    hostName = getCommunicationAddresses().get(0).toString();
+        if (hostName == null) {
+            synchronized (lock) {
+                // Work around https://bugs.openjdk.java.net/browse/JDK-8143378 on macOS
+                // See also https://stackoverflow.com/a/39698914
+                if (hostName == null && OperatingSystem.current() == OperatingSystem.MAC_OS) {
+                    ProcessBuilder builder = new ProcessBuilder("hostname");
+                    try {
+                        Process process = builder.start();
+                        int result = process.waitFor();
+                        if (result == 0) {
+                            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                            try {
+                                String line = reader.readLine();
+                                if (line != null) {
+                                    hostName = line;
+                                }
+                            } finally {
+                                reader.close();
+                            }
+                        }
+                    } catch (Exception e) {
+                        logger.debug("Couldn't resolve hostname by running hostname, falling back to using JDK", e);
+                    }
+                }
+
+                if (hostName == null) {
+                    try {
+                        hostName = InetAddress.getLocalHost().getHostName();
+                    } catch (UnknownHostException e) {
+                        hostName = getCommunicationAddresses().get(0).toString();
+                    }
                 }
             }
-            return hostName;
         }
+        return hostName;
     }
 
     /**

--- a/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/package-info.java
+++ b/subprojects/messaging/src/main/java/org/gradle/internal/remote/internal/inet/package-info.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@NonNullApi
+package org.gradle.internal.remote.internal.inet;
+
+import org.gradle.api.NonNullApi;

